.github/CODING_GUIDELINES.md

```diff
@@ -252,3 +253,109 @@ public class ThisIsASampleClassTest()
+
+```java
+
+<type>(<scope>): <subject>
+<BLANK LINE>
+<body>
+<BLANK LINE>
+<footer>
+
+```
+The header is mandatory and the body and footer are optional.
+
+Any line of the commit message cannot be longer 100 characters! This allows the message to be easier to read on GitHub as well as in various git tools.
+
+Footer should contain a closing reference to an issue if any.
+
+**Revert**
+
+If the commit reverts a previous commit, it should begin with revert:, followed by the header of the reverted commit. In the body it should say: This reverts commit <hash>., where the hash is the SHA of the commit being reverted.
+
+**Squash**
+
+Split your commits on the basis of the package you are providing code to. 
+
+Squash all your commits per scope (i.e package). It is also important to rebase your commits on master.
+
+**Type**
+
+Must be one of the following:
+
+* build: Changes that affect the build system (eg POM changes)
+* docs: Documentation only changes (eg ReadMe)
+* feat: A new feature
+* fix: A bug fix
+* perf: A code change that improves performance
+* refactor: A code change that neither fixes a bug nor adds a feature
+* style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
+* test: Adding missing tests or correcting existing tests
+
+**Scope**
+
+The scope should be the name of the maven package affected as perceived by person reading changelog generated from commit messages.
+
+The following is the list of supported scopes:
+
+* iot-dev
+* iot-dev-samples
+* iot-service
+* iot-service-samples
+* deps
+* prov-dev
+* prov-service
+* prov-samples
+* security-provider
+* tpm-provider
+* tpm-emulator
+* x509-provider
+* dice-emulator
+* dice-provider
+
+**Subject**
+
+The subject contains succinct description of the change:
+
+* use the imperative, present tense: "change" not "changed" nor "changes"
+* no dot (.) at the end
+
+**Body**
+
+Just as in the subject, use the imperative, present tense: 
+
+"change" not "changed" nor "changes". The body should include the motivation for the change and contrast this with previous behavior.
+
+**Footer**
+
+The footer should contain any information about Breaking Changes and is also the place to reference GitHub issues that this commit Closes.
+
+Breaking Changes should start with the word BREAKING CHANGE: with a space or two newlines. The rest of the commit message is then used for this.
+
+***Example commit messages***
+
+Good commit messages looks like below :
+
+* fix(iot-dev): Fix failure in reconnection
+
+    Fix failure in MQTT reconnection when token expires.
+
+    Github issue (fix#123)
+* feat(iot-service): Add continuation token to Query
+* docs: Update readme to reflect provisiong client
+
+Bad commit messages looks like below :
+
+* fix(iot-dev): small fix
+
+    I was trying to reconnect and network dropped. Fixing such random failures
```

Comment:
>Fixing [](start = 51, length = 6)

Shouldn't be "fix" if I follow the rule above for imperative?

Reply:
This is to show how a bad commit message looks like. Hence "Fixing" is used on purpose.